SonarSource/sonarlint-visualstudio
# SonarQube client

What's changed:

- the requests are no longer in a single class, but are in separate classes for each request and version
- the request/response messages are built into the request classes because they can be versioned too (properties, structure, etc. could change)
- each request class is registered into the RequestFactory, which takes care about versioning and what request to execute, e.g. no need for if-s in the code. If you want to - to handle a breaking change, either create a new request, or override an existing.
- the requests return models, each request handles errors and http response codes (because this could also change between the api versions)
- I did not write the task cancellation tests, nor tests that ensure we throw when we are not connected and call a service method; not sure if there is big benefit from them, but should be easy enough

New functionality:

- latest API changes are applied (see last few commits)

What could be improved:

- SonarQubeService.GetQualityProfileAsync is not good, it could be moved to a single Request class that does everything, or better yet - create a CompositeRequest class that accepts several requests and executes them. 

TODO (documentation)

- write a readme.txt that explains how to create a new request, how to handle a breaking change, etc.

TODO (test plan):

- connect to SonarQube
	- invalid user/pass
	- invalid url
	- correct user/pass and url

- connect to SonarQube
	- with SonarC# installed
	- without SonarC# installed

- connect to SonarQube
	- with existing projects
		- quality profile
			- with rules
			- without rules
	- without projects

- analyze a project with suppressed issues (suppressed in SQ should be hidden in VS)

- receive notification event

All connected servers:
- should be able to show project dashboard in the VS Team Explorer window

Repeat the above steps with:
- SQ 5.6.*
- SQ 6.7.*
- SQ 7.0
- SQ 7.1
- SQ 7.2 (when available)
- SonarCloud

Repeat the above steps with:
- VS 2015
- VS 2017
